mtgskynote/skynote-node
# [ENG-44]: Updated profile page with ability to change user password

**CHANGES:**
- Created NewAlert Component based on MUI that can be easily used to trigger alert messages in the UI. 
- Register.js change remains for testing but just be commented out before merging.
- Changed UI of the profile page to use tailwind and added features (instrument view, changing password, etc.) (deleted CSS file used previously for UI)
- Created new controller and routes for profile.js (profileController and profileRoutes) that will handle all cases where profile data is changed, moved updateProfileData from authController to profileController and the routing for it from authRoutes to profileRoutes. 
- Updated server.js to work with new routing.


**TESTING PROCEDURE:**

_NOTE_: it is good to test this procedure on either a new user or an existing one. However, due to bugs that need fixing, ensure that the user you use to test this has a JWT token (has been logged in in the last 2 days) and if you create a user, you also need to log in and out before you can access the profile page. These bugs have been reported and will be fixed in future iterations.

1. Basic functionality of the new Profile page
- [ ] Pull in the changes and navigate to the profile page.
- [ ] Observe the changes to the UI. The name fields and email field should contain the user data from your user in the database (check this using Compass). If some data doesn't exist (e.g. last name in the database is set to lastName), the field should be blank. (Currently, if the user edits the data, they will have to provide a last name). 
- [ ] The instrument that you have as your instrument in the database should display green while the other is grey.
- [ ] This can be tested by modifying the string in the 'instrument' field of a user in the database and observing that the colouring of the two circles changes accordingly.
- [ ]  Click edit profile and observe that a password field is displayed. This cannot be edited, but you should observe a change button next to it. 
- [ ] Also, at the bottom of the profile form, the buttons should now be "save changes" and "cancel". 
- [ ] Click the "Change" button and observe that two new fields are displayed for setting a new password. 
- [ ] Observe that the "Change" button next to the original password field is now red and says "cancel".
- [ ] Click either "cancel" button and Observe that the original state of the page is returned. 

2. Changing your user info (not password)
- [ ] Choose to edit your profile.
- [ ] Modify one of the fields that is not for the password change so it is empty.
- [ ] Press 'Save Changes' and observe a popup on the empty field that requests that it be filled.
- [ ] Change the field into whatever you want
- [ ] Press the cancel button. 
- [ ] Observe that the page returns to its original state and whatever changes you made to the field are no longer present.
- [ ] Choose to edit the information again. 
- [ ] Change a field that you want (e.g. one of the names)
- [ ] Press 'Save Changes'
- [ ] Observe an alert at the top of the profile page for 2 seconds indicating success of changing the information. 
- [ ] Go to Compass to observe that the information you changed did change in the database. 
- [ ] Observe that the profile page has returned to its initial state, but now the changes you made are present in the correct field.
- [ ] Try changing your email without an @ sign and observe that a popup shows up indicating that you must have an @ sign in the email. (note: Further iterations of user handling of app in general should include email verification etc., but this is not implemented here yet)
- [ ] Change your email to a valid email address and observe that changes are successful both by alert and in Compass. 
- [ ] Log out and log in using the new email. 
- [ ] Observe that login is successful.
- [ ] (Change your email back if you wish)

3. Changing your password
- [ ] Choose to edit your profile.
- [ ] Choose to change your password.
- [ ] Type two different things in the two password fields (mismatched passwords)
- [ ] Observe an alert come up for 2 seconds indicating that the passwords do not match.
- [ ] Click cancel (either one). 
- [ ] Navigate to change your password again.
- [ ] Observe that the two fields have been reset to being empty.
- [ ] Type the same new password into the two password fields (matching password).
- [ ] Press save. 
- [ ] Observe an alert indicating a successful password change. 
- [ ] Observe that the profile page has been reset to its original state. 
- [ ] Log out of the platform.
- [ ] Log back in using the new password you created.
- [ ] Observe that login is successful.
- [ ] (Change your password back if you wish)

3. Error in updating info 
- [ ] Choose to edit your profile.
- [ ] Navigate to the developer tools --> applications 
- [ ] Delete the token in local storage.
- [ ] Try to change your profile details. 
- [ ] Observe an alert that there is a network error.
- [ ] Log out and log back in.


**BEFORE MERGE:**
- [ ] Commnet out changed line in Register.js
